### PR TITLE
Importable HTML templates for forms

### PR DESCRIPTION
**Pull request by rask**
As per #89 this PR implements importable HTML templates for forms.

Work in progress as of now and I would like feedback on the direction I'm going towards.

### CODE REVIEW

**Comment by tnottu** [comment]
I suggest that the `wplf_import_html_template` filter would return an html string instead of a file path. Some people may want to import a form, but not from an actual html file. In my case, I would like to render a form using Twig templates.

From a quick look, this change wouldn't be a big code change. Seems that just the `file_get_contents` needs to be done in the filter callback instead of `override_form_template`.

For example:
```
add_filter( 'wplf_import_html_template', function ($template, $form_id) {
    $html_form_id = 123;
    $twig_form_id = 456;

    if ($form_id = $html_form_id) {
        return file_get_contents('/path/to/template/file.html');
    } else if ($form_id = $twig_form_id) {
        // Render a form using Timber & Twig, or by any other means.
        return Timber::compile('my-fancy-form.twig');
    }

    return $template;
} );
```

**Comment by rask** [comment]
Good point. I will implement this right away.

One thing I want to discuss with maintainers is security: do we want to add some checks for the imported HTML to harden security or do we trust developers to not do anything silly?

**Comment by rask** [comment]
Templates are now imported as HTML strings instead of HTML file paths.

What I need to implement is a fix for the following:

1. Create a form with ID `123`,
2. Create an HTML template and register it to form with ID `123`,
3. No one opens the form `123` and presses `Save`, meaning the template is not persisted to the database.

Two options:

1. Force a save somehow, which would persist the template
2. Don't load the template from the database for overridden forms.

I think the safer bet would be to add a `the_content` filter for overridden forms to load the template instead of reading it from the database. Opinions?

**Comment by k1sul1** [comment]
You can always use `wp_insert_post`? 

I think that the form has to exist in the database for things to work properly, but the content of the database entry should be overwritten using `the_content` filter as you said.

If you update the form markup in the database every time the form is loaded through the admin, it should handle the corner case of removing an override nicely.

**Comment by rask** [comment]
@k1sul1 `wp_insert_post` would work, yes. One thing I would like to avoid is unnecessary DB reads and writes to check if the database `post_content` matches the overriding template.

I wonder how much delay a read/write check on `admin_init` for this would add.

**Comment by k1sul1** [comment]
You could use wp_options to store hashes of the form content => compare that hash to the HTML string and update the post if the hash differs. And you only need to do this when the form is loaded in the admin.

**Comment by rask** [comment]
@k1sul1 maybe even transients for faster access where applicable. Will try it out. Thanks for the pointer. :+1:

**Comment by k1sul1** [comment]
Or that. Most of the time transients live in the options anyway,  but the Transients API is a lot nicer than the Options API.

**Comment by rask** [comment]
A simple template auto-persist done. I wonder if the 8 hour transient timeout is useless or fine. Need to still test how the thing works when an override is removed, may have to refactor for that use case.

**Comment by rask** [comment]
Hmm, silly mistake. There seems to be no way of overriding a WP editor field content for a single specific editor field. Need to investigate some other solution or all editor fields on the edit page will be overridden with the HTML template.

**Comment by rask** [comment]
Feature requests, feedback, and other notes welcome. If none appear soon I'll try and do final checkups for this and remove the WIP label.

It seems like Travis is failing for general code style issues and there is an upstream fix for that somewhere?

**Comment by k1sul1** [comment]
Yeah, WordPress CS continues to change resulting in the Travis build breaking.

There's a PR by me that fixes the lint errors, but the fixes are mental: https://github.com/libreform/wp-libre-form/pull/85/commits/1e60ceb307656e0b5021cec562cb91c5ec8a685b

@anttiviljami can we lock down a WPCS version to something that isn't currently broken? Basically just change dev-master on this line: https://github.com/libreform/wp-libre-form/blob/master/.travis.yml#L18

**Comment by k1sul1** [comment]
This PR seems to be working as expected.

**Comment by rask** [comment]
I will fix the PHPCS errors and then I guess we're done here for now. I can open new PRs in case this feature needs fixes or modifications later on.

**Comment by k1sul1** [comment]
I fixed most of the lint errors, feel free to rebase into master.

**Comment by rask** [comment]
Rebased and adjusted. Feel free to merge if you're OK with it.

**Comment by k1sul1** [comment]
Everything looks okay, if no one opposes, let's do a merge soon.

File: README.md [hunk 1]
```diff
@@ -207,3 +208,51 @@ The attribute will render as is on the `<form>` element
...
+
+This plugin allows you to define HTML forms in your project source code
+and import them into the form admin for specific forms.
+
+### Creating a static HTML template
+
+The simplest way is to create a HTML5 file and read its contents. Other
+options include using Twig to render HTML templates.
+
+Remember: WPLF will insert `form` tags on its own, meaning you only have
```
Comment (k1sul1) [inline README.md]: Is this necessary? I didn't read this documentation when I tried it out and it removed the form tags for me.

File: classes/class-cpt-wplf-form.php [hunk 2]
```diff
@@ -460,6 +466,138 @@ class="code"
...
+     * @param string|null $template_content Raw HTML to import for a form.
+     * @param int $form_id Form ID (WP_Post ID) to import template for.
+     */
+    $template_content = apply_filters( 'wplf_import_html_template', null, $form_id );
+
+    if ( $template_content === null ) {
+      return;
+    }
+
+    // Clear unwanted form tags. WPLF will insert those by itself when rendering a form.
```
Comment (k1sul1) [inline classes/class-cpt-wplf-form.php]: Nice addition! :+1:

File: classes/class-cpt-wplf-form.php [hunk 3]
```diff
@@ -460,6 +466,138 @@ class="code"
...
+   * Check if we need to auto-persist the form template override into WP database.
+   *
+   * @param string $template Template to maybe persist.
+   * @param int $form_id Form ID to persist template for.
+   * @param bool $force Force a persist even though not required?
+   *
+   * @return void
+   */
+  protected function maybe_persist_override_template( $template, $form_id, $force = false ) {
+    $hash_transient = 'form_tmpl_hash_' . $form_id;
```
Comment (k1sul1) [inline classes/class-cpt-wplf-form.php]: Maybe we should prefix this one with wplf_, just to be sure.

Comment (rask) [reply classes/class-cpt-wplf-form.php]: Good point, will get that into place.

File: README.md (hunk 1 above)
Comment (rask) [reply README.md]: Some people might have the idea of adding their own attributes to the `form` tag and they will get mad when they don't appear in the form for no clear reason. I would rather have that documented as the import feature is aimed towards developers who might assume the `form` tags are imported as is.

Comment (k1sul1) [reply README.md]: Makes sense. If they want to have custom attributes, they should add those to the shortcode used to print the form.

**Comment by k1sul1** [comment]
Nope. We can always revert if there's something wrong.

ship it

**Comment by rask** [comment]
Awesome, let me know if there are any issues with this so I can help where needed. :+1:

**Pull request merged**